### ISSUE TITLE
Can this loss be used for multi-label classification？

### ISSUE THREAD

**Issue by MOMOANNIE** [LIVIAETS/boundary-loss#47]
Hello, thanks for sharing your code. I want to know if this boundary loss can be used for multi-label classification? I now want to segment the lung lobes and divide the lung lobes into 5 parts. In the segmentation result, the segmentation boundary is a little bit different from the label boundary. I want the segmentation boundary to be as close to the label boundary as possible. May I use the boundary loss you proposed for multi-labell classification improved?

**Comment by HKervadec** [comment]
Hi, and thanks for the interest.

Do you mean multi-class _segmentation_ rather than _classification_? 

If yes, then it is possible; even simpler to make work than its binary counterpart, see the comparison below:
![multiclass](https://raw.githubusercontent.com/LIVIAETS/boundary-loss/master/resources/acdc_bl.png)

The parameters of the loss have to changed to something like:
```python
boundary_loss = BoundaryLoss(idc=[0, 1, 2, 3, 4])
```

In the makefile, that gives something as (for the case of ACDC):
```makefile
$(RD)/boundary: OPT = --losses="[('BoundaryLoss', {'idc': [0, 1, 2, 3]}, None, None, None, 1)]"
$(RD)/boundary: data/ACDC-2D/train/gt data/ACDC-2D-BL/val/gt
$(RD)/boundary: DATA = --folders="$(B_DATA)+[('gt', dist_map_transform, False)]"
```
I've found out that the boundary loss is able to work alone in multi-class settings, so you could start by testing than. Then you could try a more complex setting, e,g.:


For a mix of cross-entropy and boundary loss it could be (with a shifting weight strategy between the two losses:
```makefile
$(RD)/cross_entropy_surface_steal/: OPT = --losses="[('CrossEntropy', {'idc': [0, 1, 2, 3]}, None, None, None, 1), \
	('SurfaceLoss', {'idc': [0, 1, 2, 3]}, None, None, None, 0.01)]" \
	--scheduler=StealWeight --scheduler_params="{'to_steal': 0.01}"
$(RD)/cross_entropy_surface_steal/: data/ACDC-2D/train/gt data/ACDC-2D-BL/val/gt
$(RD)/cross_entropy_surface_steal/: DATA = --folders="$(B_DATA)+[('gt', gt_transform, True), \
	('gt', dist_map_transform, False)]"
```


Let me know, feel free to show an example of your task/dataset,

Hoel

**Comment by MOMOANNIE** [comment]
> Do you mean multi-class segmentation rather than classification?

Yes, I’m talking about segmentation. I want to divide the lung lobes into 5 parts. The figure below is the prediction result of my current model. My loss function uses Focal loss. From the current prediction result, you can see that it is compared with the label，the boundary still has a certain distance. I want to try to combine the boundary loss you proposed with the focal loss, whether it will make the boundary of the prediction result coincide with the boundary of the label as much as possible？
![image](https://user-images.githubusercontent.com/26375536/134790649-856c561a-c804-4f3f-a4a4-d0d1f3f6fc40.png)

**Comment by HKervadec** [comment]
> I want to try to combine the boundary loss you proposed with the focal loss, whether it will make the boundary of the prediction result coincide with the boundary of the label as much as possible？

It might help, but you will have to tune the balance between all losses, including (but not limited):
* loss (initial) weights;
* shifting the weights over time;
* supervising only a subset of the classes;
* [normalizing the distance map](https://github.com/LIVIAETS/boundary-loss/issues/14#issuecomment-816923763) or not.


Let me know how that goes.

Hoel

**Comment by MOMOANNIE** [comment]
> > I want to try to combine the boundary loss you proposed with the focal loss, whether it will make the boundary of the prediction result coincide with the boundary of the label as much as possible？
> 
> It might help, but you will have to tune the balance between all losses, including (but not limited):
> 
> * loss (initial) weights;
> * shifting the weights over time;
> * supervising only a subset of the classes;
> * [normalizing the distance map](https://github.com/LIVIAETS/boundary-loss/issues/14#issuecomment-816923763) or not.
> 
> Let me know how that goes.
> 
> Hoel

Sorry Hoel, I just saw your reply. I would like to ask if I want to add the boundary loss to the 3D image, how should I calculate the 3D distance map in the first step?
![11](https://user-images.githubusercontent.com/26375536/217987473-b1c0a453-cbc3-48ab-82b1-434aff581ca4.png)
